system/syscalls.cpp

```diff
@@ -199,10 +199,21 @@ static long mmap_new(long length)
 		set_errno(ENOMEM);
 		return -1;
 	}
-	char* ret = mmapEnd;
-	mmapEnd += length;
+	Page* newPage = reinterpret_cast<Page*>(mmapEnd);
+	mmapEnd += res;
 
-	return reinterpret_cast<long>(ret);
+	// Insert the new page into the free list to possibly merge with an existing one.
+	newPage->init(res);
+	freePages.insert(newPage);
+
+	p = freePages.lower_bound(length);
+	assert(p != freePages.end());
+	freePages.remove(p);
+	Page* rest = p->split(length);
+	if (rest)
+		freePages.insert(rest);
```

Comment:
I don't see how this logic is connected to page size. Maybe the commit name should be changed or some context added via comments

Reply:
Previously the entire 64k page was returned, but now we only return what was requested (aligned to 4k), the rest will be put in the free lists.
I can leave some comments to clarify the code.